proxy/src/main/java/net/md_5/bungee/util/ChatComponentTransformer.java

```diff
@@ -0,0 +1,113 @@
+package net.md_5.bungee.util;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import net.md_5.bungee.api.chat.BaseComponent;
+import net.md_5.bungee.api.chat.ScoreComponent;
+import net.md_5.bungee.api.chat.TextComponent;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.score.Score;
+import net.md_5.bungee.api.score.Scoreboard;
+
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * This class transforms chat components by attempting to replace transformable fields with the appropriate value.<br>
+ * ScoreComponents are transformed by replacing their {@link ScoreComponent#getName()}} into the matching entity's name
+ * as well as replacing the {@link ScoreComponent#getValue()} with the matching value in the {@link net.md_5.bungee.api.score.Scoreboard}
+ * if and only if the {@link ScoreComponent#getValue()} is not present.
+ */
+public final class ChatComponentTransformer
+{
+    private static final ChatComponentTransformer INSTANCE = new ChatComponentTransformer();
+    /**
+     * The Pattern to match entity selectors.
+     */
+    private static final Pattern SELECTOR_PATTERN = Pattern.compile( "^@([pares])(?:\\[([^ ]*)\\])?$" );
+
+    private ChatComponentTransformer() {}
+
+    public static ChatComponentTransformer getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Transform a set of components, and attempt to transform the transformable fields.
+     * Entity selectors <b>cannot</b> be evaluated.
+     * This will recursively search for all extra components (see {@link BaseComponent#getExtra()}).
+     *
+     * @param component the component to transform
+     * @return the transformed component, or an array containing a single empty TextComponent if the components are null or empty
+     * @throws IllegalArgumentException if an entity selector pattern is present
+     */
+    public BaseComponent[] transform(Scoreboard scoreboard, ProxiedPlayer player, BaseComponent... component)
+    {
+        if( component == null || component.length < 1 )
+        {
+            return new BaseComponent[] { new TextComponent( "" ) };
+        }
+
+        for (BaseComponent root : component )
+        {
+            if( root.getExtra() != null && !root.getExtra().isEmpty() )
+            {
+                List<BaseComponent> list = Lists.newArrayList( transform( scoreboard, player, root.getExtra().toArray( new BaseComponent[]{} ) ) );
+                root.setExtra( list );
+            }
+            transformComponent(scoreboard, player, root);
+        }
+        return component;
+    }
+
+    private void transformComponent(Scoreboard scoreboard, ProxiedPlayer player, BaseComponent component)
```

Comment:
I only see this referenced once, you could probably remove it to make source easier to read.

Reply:
Yes, I agree. Thanks for the advice <3. I dropped that method as well as the extraneous Scoreboard argument since you can get the Scoreboard from the ProxiedPlayer now.